map/client/client.go

```diff
@@ -0,0 +1,335 @@
+	if len(descs) == 0 {
+		return nil, errors.New("No descriptors")
+	}
+	slot := int(binary.LittleEndian.Uint64(ID[:8])) % len(descs)
+	// sort the descs and return the chosen one
+	desc := deterministicSelect(descs, slot)
+	return &mapStorage{name: desc.Name, provider: desc.Provider, id: ID}, nil
+}
+
+// Put places a value into the store
+func (c *Client) Put(ID common.MessageID, signature, payload []byte) error {
+	if !ID.WritePk().Verify(signature, payload) {
+		return errors.New("signature does not verify Write")
+	}
+
+	loc, err := c.GetStorageProvider(ID)
+	if err != nil {
+		return err
+	}
+	b := common.MapRequest{ID: ID, Signature: signature, Payload: payload}
+	// XXX: ideally we limit the number of retries
+	// so that it doesn't keep trying to deliver to a stale/missing service forever...
+	serialized, err := cbor.Marshal(b)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.Session.SendUnreliableMessage(loc.Name(), loc.Provider(), serialized)
+	// XXX: do we need to track msgId and see if it was delivered or not ???
+	return err
+}
+
+// PayloadSize returns the size of the user payload
+func (c *Client) PayloadSize() int {
+	geo := c.Session.SphinxGeometry()
+	return geo.UserForwardPayloadLength - cborFrameOverhead - eddsa.SignatureSize
+}
+
+// Get requests ID from the chosen storage node and returns a payload or error
+func (c *Client) Get(ID common.MessageID, signature []byte) ([]byte, error) {
+
+	if !ID.ReadPk().Verify(signature, ID[:]) {
+		return nil, errors.New("signature does not verify Read")
+	}
+	loc, err := c.GetStorageProvider(ID)
+	if err != nil {
+		return nil, err
+	}
+	b := &common.MapRequest{ID: ID, Signature: signature}
+	serialized, err := cbor.Marshal(b)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := c.Session.BlockingSendUnreliableMessage(loc.Name(), loc.Provider(), serialized)
+	if err != nil {
+		return nil, err
+	}
+	// unwrap the response and return the payload
+	resp := &common.MapResponse{}
+	err = cbor.Unmarshal(r, resp)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Status == common.StatusNotFound {
+		return nil, ErrStatusNotFound
+	}
+	return resp.Payload, nil
+}
+
+// RWClient has both Get and Put
+type RWClient interface {
+	Put(addr []byte, payload []byte) error
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// ROClient only has Get
+type ROClient interface {
+	Get(addr []byte) ([]byte, error)
+	PayloadSize() int
+}
+
+// WOClient only has Put
+type WOClient interface {
+	Put(addr []byte, payload []byte) error
+	PayloadSize() int
+}
+
+// rwMap implements ReadWriteTranport using a ReadWriteCap and reference to map client
+// The idea is that several streams can share map/client and client/session but
+// have different capabilities for the underlying Streams transported by map/client
+type rwMap struct {
+	c     *Client
+	rwCap common.ReadWriteCap
+}
+
+// Get implements RWClient.Get
+func (r *rwMap) Get(addr []byte) ([]byte, error) {
```

Comment:
ideally we'd be able to test the full functionality of a capability without any client IO

Reply:
I agree that https://github.com/katzenpost/katzenpost/blob/add_reliable_streams/map/common/cap_test.go is incomplete and will add additional tests for each capability type.